cylc/flow/task_job_mgr.py

```diff
@@ -1181,6 +1191,14 @@ def _prep_submit_task_job(
                 itask.summary['platforms_used'][itask.submit_num] = ''
                 # Retry delays, needed for the try_num
                 self._create_job_log_path(workflow, itask)
+                if isinstance(exc, NoPlatformsError):
+                    # Todo = need to clear all hosts from all platforms
+                    # in group.
+                    self.bad_hosts -= exc.hosts_consumed
+                    self._set_retry_timers(itask, rtconfig)
+                    self._prep_submit_task_job_error(
+                        workflow, itask, '(no platforms available)', exc)
+                    return exc
```

Comment:
Do we actually *need* to separate "bad tasks" from "bad platform group tasks" beyond this point?

I think we should be able to make this work even returning `False` at this point?

Reply:
Yes, we do. We don't want to return them from `submit_task_jobs` in the `bad_tasks` list, but in the `done_tasks` - If they are added to the bad tasks the retry doesn't work.

Reply:
I struggled to make this work in practice - I'll have another look and see if I remember why...

Reply:
It looks like this distinction is non-functional, but is probably used in the tests?